src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs

```diff
@@ -0,0 +1,208 @@
+				if( !ProhibitedAssemblyCategories.Contains( category ) ) {
+					return;
+				}
+
+				context.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.NUnitCategory,
+					attribute.ApplicationSyntaxReference.GetSyntax( context.CancellationToken ).GetLocation(),
+					$"Assemblies cannot be categorized as any of [{string.Join( ", ", ProhibitedAssemblyCategories )}], but saw '{category}'."
+				) );
+			} );
+		}
+
+		private static void AnalyzeMethod(
+			SyntaxNodeAnalysisContext context,
+			NUnitTypes types,
+			MethodDeclarationSyntax syntax
+		) {
+			SemanticModel model = context.SemanticModel;
+
+			
+			IMethodSymbol method = model.GetDeclaredSymbol( syntax, context.CancellationToken );
+			if( method == null ) {
+				return;
+			}
+
+			if( !IsTestMethod( types, method ) ) {
+				return;
+			}
+
+			ImmutableSortedSet<string> categories = GatherTestMethodCategories( types, method );
+
+			if( categories.Overlaps( RequiredCategories ) ) {
+				return;
+			}
+
+			context.ReportDiagnostic( Diagnostic.Create(
+				Diagnostics.NUnitCategory,
+				syntax.Identifier.GetLocation(),
+				$"Test must be categorized as one of [{string.Join( ", ", RequiredCategories )}], but saw [{string.Join( ", ", categories )}]. See http://docs.dev.d2l/index.php/Test_Categories."
+			) );
+		}
+
+		private static bool IsTestMethod(
+			NUnitTypes types,
+			IMethodSymbol method
+		) {
+			foreach( AttributeData attribute in method.GetAttributes() ) {
+				INamedTypeSymbol attributeType = attribute.AttributeClass;
+				if( types.TestAttributes.Contains( attributeType ) ) {
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		private static ImmutableSortedSet<string> GatherTestMethodCategories(
+			NUnitTypes types,
+			IMethodSymbol method
+		) {
+			ImmutableSortedSet<string> methodCategories = GetCategories( types, method );
+			ImmutableSortedSet<string> fixtureCategories = GetCategories( types, method.ContainingType );
+			ImmutableSortedSet<string> assemblyCategories = GetCategories( types, method.ContainingAssembly );
+
+			ImmutableSortedSet<string> categories =
+				methodCategories
+					.Union( fixtureCategories )
+					.Union( assemblyCategories );
+
+			return categories;
+		}
+
+		private static ImmutableSortedSet<string> GetCategories(
+			NUnitTypes types,
+			ISymbol symbol
+		) {
+			var categories = ImmutableSortedSet.CreateBuilder<string>( StringComparer.OrdinalIgnoreCase );
+
+			VisitCategories( types, symbol, ( category, _ ) => categories.Add( category ) );
+
+			return categories.ToImmutable();
+		}
+
+		private static void VisitCategories(
+			NUnitTypes types,
+			ISymbol symbol,
+			Action<string, AttributeData> visitor
+		) {
+			foreach( AttributeData attribute in symbol.GetAttributes() ) {
+				INamedTypeSymbol attributeType = attribute.AttributeClass;
+				if( types.CategoryAttribute == attributeType ) {
+					visitor( attribute.ConstructorArguments[0].Value as string, attribute );

```

Comment:
Hm, a bit of an assumption here?

Reply:
As opposed to `GetConstantValue`? Yeah that could probable make sense

Reply:
although the constant value might already be resolved here, I think. `ConstructorArguments` is `TypedConstant[]`.

So I am assuming it's a string yeah, but, the constructor is `( string name )`

Reply:
Ah, yeah, attributes. I was wondering how we knew it was a constant.
It's worth double checking manually/in a test that it'll work if someone uses a `const` for some reason.